src/agent/onefuzz-task/src/tasks/coverage/dotnet.rs

```diff
@@ -0,0 +1,372 @@
+use anyhow::{Context, Result};
+use async_trait::async_trait;
+use onefuzz::{
+    expand::{Expand, PlaceHolder},
+    monitor::DirectoryMonitor,
+    syncdir::SyncedDir,
+};
+use reqwest::Url;
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+    process::Stdio,
+    time::Duration,
+};
+use std::{env, process::ExitStatus};
+use storage_queue::{Message, QueueClient};
+use tokio::{fs, process::Command, time::timeout};
+use tokio_stream::wrappers::ReadDirStream;
+use uuid::Uuid;
+
+use crate::tasks::{
+    config::CommonConfig,
+    coverage::COBERTURA_COVERAGE_FILE,
+    generic::input_poller::{CallbackImpl, InputPoller, Processor},
+    heartbeat::{HeartbeatSender, TaskHeartbeatClient},
+};
+
+const MAX_COVERAGE_RECORDING_ATTEMPTS: usize = 2;
+const DEFAULT_TARGET_TIMEOUT: Duration = Duration::from_secs(5);
+
+#[derive(Debug, Deserialize)]
+pub struct Config {
+    pub target_exe: PathBuf,
+    pub target_env: HashMap<String, String>,
+    pub target_options: Vec<String>,
+    pub target_timeout: Option<u64>,
+
+    pub input_queue: Option<QueueClient>,
+    pub readonly_inputs: Vec<SyncedDir>,
+    pub coverage: SyncedDir,
+
+    #[serde(flatten)]
+    pub common: CommonConfig,
+}
+
+impl Config {
+    pub fn timeout(&self) -> Duration {
+        self.target_timeout
+            .map(Duration::from_secs)
+            .unwrap_or(DEFAULT_TARGET_TIMEOUT)
+    }
+}
+
+pub struct DotnetCoverageTask {
+    config: Config,
+    poller: InputPoller<Message>,
+}
+
+impl DotnetCoverageTask {
+    pub fn new(config: Config) -> Self {
+        let poller = InputPoller::new("dotnet_coverage");
+        Self { config, poller }
+    }
+
+    pub async fn run(&mut self) -> Result<()> {
+        info!("starting dotnet_coverage task");
+        self.config.coverage.init_pull().await?;
+
+        let heartbeat = self.config.common.init_heartbeat(None).await?;
+        let mut context = TaskContext::new(&self.config, heartbeat);
+
+        if !context.uses_input() {
+            bail!("input is not specified on the command line or arguments for the target");
+        }
+
+        context.heartbeat.alive();
+
+        let coverage_local_path = self.config.coverage.local_path.canonicalize()?;
+        let intermediate_files_path = intermediate_coverage_files_path(&coverage_local_path)?;
+        fs::create_dir_all(&intermediate_files_path).await?;
+        let timeout = self.config.timeout();
+        let coverage_dir = self.config.coverage.clone();
+
+        tokio::spawn(async move {
+            info!("Starting directory monitor");
+            let mut monitor = DirectoryMonitor::new(intermediate_files_path)
+                .await
+                .unwrap();
+            info!("Started directory monitor, waiting for files");
+            while (monitor.next_file().await.unwrap()).is_some() {
+                info!("Found intermediate coverage file");
```

Comment:
You're right it's more for debugging (did we see the file and fail to process or not see a file at all?) I kept the `info!("Merged and synced coverage");` log as info since I think there's still value in logging progress and moved this one to debug.